# Import Libraries

In [1]:
import pandas as pd
import numpy as np
import os

In [2]:
# Turn your project folder path into a string

path = r'/Users/peterreadman/Desktop/Python Projects/CareerFoundry/04-2020 Instacart Basket Analysis/'

# Exercise read-through

In [3]:
# import orders_products_merged.pkl dataframe from the pkl file

ords_prods_merged = pd.read_pickle(os.path.join(path, '02 Data', '02b Prepared Data', 'orders_products_merged.pkl'))

In [4]:
# Create a subset

df = ords_prods_merged[:1000000]

In [26]:
df.shape

(1000000, 17)

In [27]:
# define price_label function to return low- mid- and high-range products by price

def price_label(row):

    if row['prices'] <= 5:
        return 'Low-range product'
    elif (row['prices'] > 5) and (row['prices'] <= 15):
        return 'Mid-range product'
    elif row['prices'] > 15:
        return 'High-range'
    else: return 'Not enough data'

In [28]:
# Add new column 'price_range' to df to store results of price_label()

df['price_range'] = df.apply(price_label, axis=1)

/var/folders/h0/wxw40lrj64j6l30lq749x76r0000gn/T/ipykernel_14011/2188526707.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['price_range'] = df.apply(price_label, axis=1)


In [29]:
# Check the value count in the new column (count of 'Mid-range, Low-range, High-range products)')

df['price_range'].value_counts(dropna = False)

Mid-range product    756450
Low-range product    243550
Name: price_range, dtype: int64

In [30]:
# Check the max price in 'prices' column (should be < 15)
df['prices'].max()

14.8

# If-Statements with the loc() Function

In [10]:
df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'

/var/folders/h0/wxw40lrj64j6l30lq749x76r0000gn/T/ipykernel_14011/1169838859.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'


In [11]:
df.loc[(df['prices'] <= 15) & (df['prices'] > 5), 'price_range_loc'] = 'Mid-range product'

In [12]:
df.loc[df['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [13]:
df['price_range_loc'].value_counts(dropna = False)

Mid-range product    756450
Low-range product    243550
Name: price_range_loc, dtype: int64

# Now conduct filtering on entire dataframe

Now that you’ve seen how much faster loc() works, you can try repeating the process again—this time, on your entire dataframe as opposed to the subset. The only difference in the code is the use of 'ords_prods_merged' instead of df:

In [14]:
ords_prods_merged.loc[ords_prods_merged['prices'] > 15, 'price_range_loc'] = "High-range product"

In [15]:
ords_prods_merged.loc[(ords_prods_merged['prices'] <= 15) & (ords_prods_merged['prices'] > 5), 'price_range_loc'] = "Mid-range product"

In [16]:
ords_prods_merged.loc[ords_prods_merged['prices'] <= 5, 'price_range_loc'] = "Low-range product"

In [17]:
ords_prods_merged['price_range_loc'].value_counts(dropna = False)

Mid-range product     21860860
Low-range product     10126321
High-range product      417678
Name: price_range_loc, dtype: int64

In [18]:
ords_prods_merged.shape

(32404859, 16)

# If-Statements with For-Loops

Create a new column in your ords_prods_merge dataframe that summarizes how busy each day of the week is

To start, you need to know on which day most orders take place. You can find this out by printing the frequency of the “orders_day_of_the_week” column, which would look like this:

In [19]:
# print the frequency of 'orders_day_of_the_week'

ords_prods_merged['orders_day_of_the_week'].value_counts(dropna = False)

0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: orders_day_of_the_week, dtype: int64

In [20]:
# 0 = Saturday
# 1 = Sunday
# 2 = Monday
# 3 = Tuesday
# 4 = Wednesday
# 5 = Thursday
# 6 = Friday

# create an empty list, result
result = []

# for loop

for value in ords_prods_merged["orders_day_of_the_week"]:
    if value == 0:
        result.append("Busiest day")
    elif value == 4:
        result.append("Least busy")
    else:
        result.append("Regularly busy")
    

In [21]:
result

['Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Busiest day',
 'Regularly busy',
 'Reg

In [22]:
# Create a new column called 'busiest_day' in 'ords_prods_merged' for result

ords_prods_merged['busiest_day'] = result

In [23]:
ords_prods_merged['busiest_day'].value_counts(dropna = False)

Regularly busy    22416875
Busiest day        6204182
Least busy         3783802
Name: busiest_day, dtype: int64

# Task

# Question 1
*If you haven’t done so already, complete the instructions in the Exercise for creating the “price_label” and “busiest_day” columns.*
</br>
</br>
**Done, as above**

# Question 2
*Suppose your clients have changed their minds about the labels you created in your “busiest_day” column. Now, they want “Busiest day” to become “Busiest days” (plural). This label should correspond with the two busiest days of the week as opposed to the single busiest day. At the same time, they’d also like to know the two slowest days. Create a new column for this using a suitable method.*

In [24]:
# Edit the for loop by adding elif statements to check for the second busiest and second least busy days

for value in ords_prods_merged["orders_day_of_the_week"]:
    if value == 0:
        result.append("Busiest days")
    elif value == 1:
        result.append("Busiest days")
    elif value == 4:
        result.append("Least busy days")
    elif value == 3:
        result.append("Least busy days")
    else:
        result.append("Regularly busy")

In [31]:
# Create a new column called 'busiest_days' (plural) in 'ords_prods_merged' for result

ords_prods_merged['busiest_days'] = result

ValueError: Length of values (64809718) does not match length of index (32404859)

In [32]:
# call result
result

['Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Busiest day',
 'Regularly busy',
 'Reg

# Question 3
*Check the values of this new column for accuracy. Note any observations in markdown format.*


In [ ]:
ords_prods_merged['busiest_days'].value_counts(dropna = False)

The total count = 32,404,859 matches the shape of the dataframe, so all values have been captured 

The **two busiest days** account for **36%** of all orders (11,864,412)

The **two least busy days** account for **23.5%** of all orders (7,624,336)



# Question 4
*When too many users make Instacart orders at the same time, the app freezes. The senior technical officer at Instacart wants you to identify the busiest hours of the day. Rather than by hour, they want periods of time labeled “Most orders,” “Average orders,” and “Fewest orders.” Create a new column containing these labels called “busiest_period_of_day.”*

In [ ]:
# Check frequency of 'order_hour_of_day' column
# Midnight = 0

ords_prods_merged['order_hour_of_day'].value_counts(dropna = False)

**10am** is the **busiest hour**
</br>
**3am** is the **least busy hour** *etc.*

In [ ]:
# Periods of time:
# "Most orders" = top 6 hours are 10:00 – 15:00 (6 hours total)
# "Fewest orders" = bottom 6 hours are 00:00 – 5:00 (6 hours total) 
# "Average orders" = others 16:00 – 23:00 (12 hours total)

result = []

for value in ords_prods_merged["order_hour_of_day"]:
    if value == 10:
        result.append("Most orders")
    elif value == 11:
        result.append("Most orders")
    elif value == 12:
        result.append("Most orders")
    elif value == 13:
        result.append("Most orders")
    elif value == 14:
        result.append("Most orders")
    elif value == 15:
        result.append("Most orders")
  
    elif value <= 5:
        result.append("Fewest orders")
    else:
        result.append("Average orders")

In [ ]:
# Check result
result

In [ ]:
# Create a new column called 'busiest_days' (plural) in 'ords_prods_merged' for result

ords_prods_merged['busiest_period_of_day'] = result

# Question 5
*Print the frequency for this new column*

In [ ]:
ords_prods_merged['busiest_period_of_day'].value_counts(dropna = False)

# Question 6
*Ensure your notebook is clean and structured and that your code is well commented*
</br>
</br>
**Hopefully it's ok**

# Question 7
*Export your dataframe as a pickle file (since you added new columns) and store it correctly in your “Prepared Data” folder*

In [ ]:
# saving as separate .pkl file 'orders_products_merged_busiest'

ords_prods_merged.to_pickle((os.path.join(path, '02 Data', '02b Prepared Data', 'orders_products_merged_busiest.pkl')))